File: src/Annora.Domain/Listings/Listing.cs
namespace Annora.Domain.Listings;

public sealed class Listing
{
    public Guid Id { get; }
    public string Title { get; }
    public string Category { get; }
    public string Description { get; }
    public decimal Price { get; }
    public string Condition { get; }
    public string PostalCode { get; }
    public string City { get; }
    public string SellerEmail { get; }
    public Guid? PrimaryImageId { get; }    
    public DateTimeOffset CreatedAt { get; }

    private Listing(
        Guid id,
        string title,
        string category,
        string description,
        decimal price,
        string condition,
        string postalCode,
        string city,
        string sellerEmail,
        Guid? primaryImageId,        
        DateTimeOffset createdAt)
    {
        Id = id;
        Title = title;
        Category = category;
        Description = description;
        Price = price;
        Condition = condition;
        PostalCode = postalCode;
        City = city;
        SellerEmail = sellerEmail;
        PrimaryImageId = primaryImageId;
        CreatedAt = createdAt;
    }

    public static Listing Create(
        string title,
        string category,
        string description,
        decimal price,
        string condition,
        string postalCode,
        string city,
        string sellerEmail,
        Guid? primaryImageId)
    {
        return new Listing(
            Guid.NewGuid(),
            title.Trim(),
            category.Trim(),
            description.Trim(),
            price,
            condition.Trim(),
            postalCode,
            city.Trim(),
            sellerEmail.Trim(),
            primaryImageId,
            DateTimeOffset.UtcNow);
    }

    public static Listing Restore(
        Guid id,
        string title,
        string category,
        string description,
        decimal price,
        string condition,
        string postalCode,
        string city,
        string sellerEmail,
        Guid? primaryImageId,
        DateTimeOffset createdAt)
    {
        return new Listing(
            id,
            title,
            category,
            description,
            price,
            condition,
            postalCode,
            city,
            sellerEmail,
            primaryImageId,
            createdAt);
    }

}

File: src/Annora.Web/js/listing.js
"use strict";

import { config } from "./config.js";
import { getListing } from "./api.js";

const loadingState = document.getElementById("loading-state");
const errorState = document.getElementById("error-state");
const errorMessage = document.getElementById("error-message");
const listingContent = document.getElementById("listing-content");

const breadcrumbTitle =
    document.getElementById("breadcrumb-title");

const titleElement = document.getElementById("title");
const categoryElement = document.getElementById("category");
const createdAtElement = document.getElementById("created-at");
const priceElement = document.getElementById("price");
const conditionElement = document.getElementById("condition");
const locationElement = document.getElementById("location");
const descriptionElement = document.getElementById("description");
const mainImage =
    document.getElementById("main-image");

const mainImagePlaceholder =
    document.getElementById(
        "main-image-placeholder"
    );

async function loadListing() {
    const parameters = new URLSearchParams(window.location.search);
    const id = parameters.get("id");

    if (!id) {
        showError("Annonce-id mangler i adressen.");
        return;
    }

    try {
        const listing = await getListing(id);

        renderListing(listing);
    }
    catch (error) {
        console.error(error);
        showError(error.message);
    }
}

function renderListing(listing) {
    document.title = `${listing.title} – Annora`;

    breadcrumbTitle.textContent = listing.title;
    titleElement.textContent = listing.title;

    categoryElement.textContent =
        getCategoryLabel(listing.category);

    createdAtElement.dateTime = listing.createdAt;
    createdAtElement.textContent =
        formatDate(listing.createdAt);

    priceElement.textContent =
        new Intl.NumberFormat("da-DK", {
            style: "currency",
            currency: "DKK",
            maximumFractionDigits: 0
        }).format(listing.price);

    conditionElement.textContent =
        getConditionLabel(listing.condition);

    locationElement.textContent =
        `${listing.postalCode} ${listing.city}`;

    descriptionElement.textContent = listing.description;

    loadingState.hidden = true;
    errorState.hidden = true;
    listingContent.hidden = false;

    if (listing.primaryImageId) {
        mainImage.src =
            `${config.apiBaseUrl}/images/` +
            encodeURIComponent(listing.primaryImageId);

        mainImage.alt = listing.title;
        mainImage.hidden = false;
        mainImagePlaceholder.hidden = true;

        mainImage.addEventListener("error", () => {
            mainImage.hidden = true;
            mainImagePlaceholder.hidden = false;
        });
    }
}

function showError(message) {
    loadingState.hidden = true;
    listingContent.hidden = true;
    errorState.hidden = false;
    errorMessage.textContent = message;
}

function getCategoryLabel(category) {
    const categories = {
        furniture: "Møbler",
        electronics: "Elektronik",
        bicycles: "Cykler",
        clothing: "Tøj",
        garden: "Have",
        tools: "Værktøj",
        books: "Bøger",
        other: "Andet"
    };

    return categories[category] ?? category;
}

function getConditionLabel(condition) {
    const conditions = {
        new: "Som ny",
        good: "God",
        used: "Brugt",
        worn: "Slidt"
    };

    return conditions[condition] ?? condition;
}

function formatDate(value) {
    return new Intl.DateTimeFormat("da-DK", {
        day: "numeric",
        month: "long",
        year: "numeric"
    }).format(new Date(value));
}

loadListing();